File: src/core/net/sanitize.test.ts
import { describe, expect, it } from 'vitest'
import { readFixtureBytes } from './__fixtures__'
import { decodeGb18030 } from './encoding/gb18030'
import { sanitizeNgaJson } from './sanitize'

const parse = (raw: string) => JSON.parse(sanitizeNgaJson(raw)) as unknown

describe('sanitizeNgaJson · API 文档 §0.6 的清洗步骤', () => {
  it('1. 剥 window.script_muti_get_var_store= 前缀', () => {
    expect(sanitizeNgaJson('window.script_muti_get_var_store={"data":1}')).toBe('{"data":1}')
  })

  it('1b. lite=htmljs 的 HTML 外壳只留 script 里的那段', () => {
    const raw =
      '<html><body><script>window.script_muti_get_var_store={"data":{"__MESSAGE":{"0":0}}};</script></body></html>'
    expect(parse(raw)).toEqual({ data: { __MESSAGE: { '0': 0 } } })
  })

  it('2. 截断 /*error fill content 尾巴', () => {
    const raw = '{"data":{"0":"ok"}}/*error fill content 这里全是垃圾'
    expect(parse(raw)).toEqual({ data: { '0': 'ok' } })
  })

  it('3. 去 /*$js$*/ 注释标记', () => {
    expect(parse('{"data":/*$js$*/{"0":1}}')).toEqual({ data: { '0': 1 } })
  })

  it('4. 修非法数字：前导 + 与前导 0 都转成字符串', () => {
    expect(parse('{"content":+123,"subject":+45,"author":0678}')).toEqual({
      content: '+123',
      subject: '+45',
      author: '0678',
    })
    expect(parse('{"a":1,"content":0123}')).toEqual({ a: 1, content: '0123' })
  })

  it('4b. 合法数字不动', () => {
    expect(parse('{"content":123,"subject":0,"author":-4}')).toEqual({
      content: 123,
      subject: 0,
      author: -4,
    })
  })

  it('5. 删坏字段 alterinfo（部分页面打不开的原因）', () => {
    expect(parse('{"pid":1,"alterinfo":"[已编辑 2 次] ","lou":3}')).toEqual({ pid: 1, lou: 3 })
  })

  it('6. 整数 key 加引号', () => {
    expect(parse('{0:"a",1:{2:"b"}}')).toEqual({ '0': 'a', '1': { '2': 'b' } })
    expect(parse('{"x":1, 12:"y"}')).toEqual({ x: 1, '12': 'y' })
  })

  it('6b. 字符串里长得像整数 key 的内容不能被改', () => {
    // 正文里出现 `{12:` 时用裸正则会把用户内容改坏，反而解析失败 → 被误判成被封
    const raw = '{"content":"看这段代码 {12:34} 还有 ,56: 这种"}'
    expect(parse(raw)).toEqual({ content: '看这段代码 {12:34} 还有 ,56: 这种' })
  })

  it('7. 字符串内的裸控制字符转义', () => {
    const raw = '{"content":"第一行\n第二行\ttab"}'
    expect(parse(raw)).toEqual({ content: '第一行\n第二行\ttab' })
  })

  it('8. 去掉赋值残留的外层括号与结尾分号', () => {
    expect(sanitizeNgaJson('window.script_muti_get_var_store=({"data":1});')).toBe('{"data":1}')
  })

  it('全部步骤叠一起也能洗成合法 JSON', () => {
    const raw =
      'window.script_muti_get_var_store=({"data":{0:{"pid":9,"alterinfo":"[edited 2] ",' +
      '"content":+7,"subject":012,"note":"裸控制符"}}});/*error fill content xxx'
    expect(parse(raw)).toEqual({
      data: { '0': { pid: 9, content: '+7', subject: '012', note: '裸控制符' } },
    })
  })
})

describe('sanitizeNgaJson · 真实抓包样本', () => {
  const cases = [
    ['notiEmpty', 'notiEmpty'],
    ['threadList', 'threadList'],
    ['readThread', 'readThread'],
    ['ucpUser', 'ucpUser'],
    ['ucpNotFound', 'ucpNotFound'],
    ['readThreadNotFound', 'readThreadNotFound'],
  ] as const

  it.each(cases)('%s 洗完是合法 JSON', (name) => {
    const text = decodeGb18030(readFixtureBytes(name))
    expect(() => JSON.parse(sanitizeNgaJson(text)) as unknown).not.toThrow()
  })

  it('read.php 的 lite=js 前缀被剥掉', () => {
    const text = decodeGb18030(readFixtureBytes('readThread'))
    expect(text.startsWith('window.script_muti_get_var_store=')).toBe(true)
    expect(sanitizeNgaJson(text).startsWith('{')).toBe(true)
  })
})

File: app/src/test/kotlin/com/chasel/ng2n/core/net/SanitizeTest.kt
package com.chasel.ng2n.core.net

import com.chasel.ng2n.golden.Goldens
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SanitizeTest {

  private val json = Json

  private fun goldenInput(name: String): String =
    Goldens.load("sanitize").first { it.name == name }.inputString()

  private fun sanitizeAndParse(raw: String): JsonObject =
    json.parseToJsonElement(sanitizeNgaJson(raw)).jsonObject

  @Test
  fun `alterinfo 保留,裸 TAB 只被转义不被删`() {
    val parsed = sanitizeAndParse("{\"pid\":1,\"alterinfo\":\"[E1748252294 0 0]\t\",\"lou\":3}")

    assertEquals("[E1748252294 0 0]\t", parsed.getValue("alterinfo").jsonPrimitive.content)
    assertEquals("1", parsed.getValue("pid").jsonPrimitive.content)
    assertEquals("3", parsed.getValue("lou").jsonPrimitive.content)
  }

  @Test
  fun `真实抓包的 alterinfo 字段洗完还在`() {
    val parsed = sanitizeAndParse(goldenInput("capture-read-thread-jsvar"))
    val floors = parsed.getValue("data").jsonObject.getValue("__R").jsonObject

    assertTrue(floors.isNotEmpty(), "抓包里应当有楼层")
    floors.forEach { (lou, floor) ->
      assertTrue(floor.jsonObject.containsKey("alterinfo"), "第 $lou 楼的 alterinfo 被吃掉了")
    }
  }

  @Test
  fun `六步叠在一起洗完是合法 JSON`() {
    val raw = "window.script_muti_get_var_store=({\"data\":{0:{\"pid\":9," +
      "\"alterinfo\":\"[E1 0 0]\t\",\"content\":+7,\"subject\":012,\"note\":\"裸控制符\"}}});" +
      "/*error fill content xxx"

    val floor = sanitizeAndParse(raw).getValue("data").jsonObject.getValue("0").jsonObject

    assertEquals("9", floor.getValue("pid").jsonPrimitive.content)
    assertEquals("[E1 0 0]\t", floor.getValue("alterinfo").jsonPrimitive.content)
    assertTrue(floor.getValue("content").jsonPrimitive.isString)
    assertEquals("+7", floor.getValue("content").jsonPrimitive.content)
    assertEquals("012", floor.getValue("subject").jsonPrimitive.content)
  }

  @Test
  fun `真实抓包洗完都是合法 JSON`() {
    val captures = Goldens.load("sanitize").filter { it.name.startsWith("capture-") }
    assertEquals(6, captures.size, "抓包样本条数变了,先确认金样本是不是重导过")

    for (case in captures) {
      val parsed = json.parseToJsonElement(sanitizeNgaJson(case.inputString()))
      assertTrue(parsed is JsonObject, "${case.name} 洗完不是 JSON 对象")
    }
  }

  @Test
  fun `read_php 的 lite=js 前缀被剥掉`() {
    val raw = goldenInput("capture-read-thread-jsvar")

    assertTrue(raw.startsWith("window.script_muti_get_var_store="), "样本本身就该带前缀")
    assertTrue(sanitizeNgaJson(raw).startsWith("{"), "剥完应当直接是对象")
  }

  @Test
  fun `字符串里长得像整数 key 的内容不能被改`() {
    val parsed = sanitizeAndParse("{\"content\":\"看这段代码 {12:34} 还有 ,56: 这种\"}")

    assertEquals("看这段代码 {12:34} 还有 ,56: 这种", parsed.getValue("content").jsonPrimitive.content)
  }

  @Test
  fun `合法数字不动`() {
    val parsed = sanitizeAndParse("{\"content\":123,\"subject\":0,\"author\":-4}")

    assertTrue(!parsed.getValue("content").jsonPrimitive.isString)
    assertEquals("123", parsed.getValue("content").jsonPrimitive.content)
    assertEquals("0", parsed.getValue("subject").jsonPrimitive.content)
    assertEquals("-4", parsed.getValue("author").jsonPrimitive.content)
  }

  @Test
  fun `去掉赋值残留的外层括号与结尾分号`() {
    assertEquals("{\"data\":1}", sanitizeNgaJson("window.script_muti_get_var_store=({\"data\":1});"))
    assertEquals("{\"data\":1}", sanitizeNgaJson("({\"data\":1}) ; ;"))
  }

  @Test
  fun `htmljs 的 HTML 外壳只留 script 里的那段`() {
    val raw = "<html><body><script>window.script_muti_get_var_store=" +
      "{\"data\":{\"__MESSAGE\":{\"0\":0}}};</script></body></html>"

    assertEquals("{\"data\":{\"__MESSAGE\":{\"0\":0}}}", sanitizeNgaJson(raw))
  }
}
